tensorflow/compiler/xla/service/all_reduce_reassociate.cc

```diff
@@ -57,16 +108,31 @@ StatusOr<bool> AllReduceReassociate::Run(
   bool changed = false;
   for (auto computation : module->computations(execution_threads)) {
     for (HloInstruction *inst : computation->MakeInstructionPostOrder()) {
+      HloInstruction *convert0 = nullptr;
+      HloInstruction *convert1 = nullptr;
       std::optional<ReductionKind> kind = MatchReductionInstruction(inst);
-      if (!kind || inst->operand(0)->opcode() != HloOpcode::kAllReduce ||
-          inst->operand(1)->opcode() != HloOpcode::kAllReduce ||
-          !inst->shape().IsArray()) {
+      if (!kind || !MatchAllreduceOperandsWithOptionalConvert(inst, &convert0,
+                                                              &convert1)) {
+        continue;
+      }
+      // Check to see if input converts are casting to different types,
+      // this is unlikely to happen but just to be safe.
+      bool should_promote_ar = convert0 || convert1;
+      if (should_promote_ar && !AreCompatibleConverts(convert0, convert1)) {
+        VLOG(2) << "Inputs have Converts of different types, skipping";
         continue;
       }
 
-      auto *ar0 = Cast<HloAllReduceInstruction>(inst->mutable_operand(0));
-      auto *ar1 = Cast<HloAllReduceInstruction>(inst->mutable_operand(1));
-      if (!AreCompatible(ar0, ar1, *kind)) {
+      HloInstruction *op_operand0 = inst->mutable_operand(0);
+      HloInstruction *op_operand1 = inst->mutable_operand(1);
+      if (should_promote_ar) {
+        if (convert0) op_operand0 = convert0->mutable_operand(0);
+        if (convert1) op_operand1 = convert1->mutable_operand(0);
+      }
+
+      auto *ar0 = Cast<HloAllReduceInstruction>(op_operand0);
+      auto *ar1 = Cast<HloAllReduceInstruction>(op_operand1);
+      if (!should_promote_ar && !AreCompatible(ar0, ar1, *kind)) {
```

Comment:
why is `!should_promote_ar` required here? Even if `should_promote_ar` is true, we still need to check the compatibility of the all-reduces. Can you update the test as well to exercise this case?

Reply:
Yea i agree that we should check compatibility for it. This is for cases that have this structure:
                      bf16-AR
                       /
                       /
f32-AR         convert(f32)
        \            /
          \         /
             add
we still reorder and promote AR to the wider type. The compatibility check will fail since the 2 ARs have different keys. Maybe in this case we just check everything else except the element type?

Reply:
Agreed, in that case we need to check everything except element type.